feat: Add ``DatabaseExecutor`` unified executor type

uuushiro opened the pull request:
- Close https://github.com/SeaQL/sea-orm/issues/2827

## Summary

- Introduces `DatabaseExecutor<'c>` enum that wraps either a `&DatabaseConnection` or `&DatabaseTransaction`
- Implements `ConnectionTrait` and `TransactionTrait` via delegation pattern
- Adds `IntoDatabaseExecutor` trait for ergonomic function signatures
- Promotes the pattern from `sea-orm-migration`'s `SchemaManagerConnection` to core

## Motivation

In real-world applications, we frequently need functions that can accept either a connection or transaction dynamically. This type solves:

- **Runtime Branching**: Deciding at runtime whether to run atomically (in a transaction) or directly
- **Test Isolation**: Reusing production repository logic in tests where transactions are rolled back
- **Safe Boundaries**: Allowing callers to manage transaction lifecycle while passing only "execution rights"

## Usage Example

```rust
use sea_orm::{DatabaseExecutor, ConnectionTrait, TransactionTrait};

async fn save_user(db: &DatabaseExecutor<'_>, name: String) -> Result<(), DbErr> {
    user::ActiveModel { name: Set(name), ..Default::default() }
        .insert(db)
        .await?;
    Ok(())
}

// Works with both connection and transaction
save_user(&DatabaseExecutor::from(&conn), "Alice".into()).await?;

let txn = conn.begin().await?;
save_user(&DatabaseExecutor::from(&txn), "Bob".into()).await?;
txn.commit().await?;
```

Huliiiiii commented:
ConnectionOrTransaction is a bit verbose. How about renaming it to DatabaseExecutor?

uuushiro commented:
@Huliiiiii 
I have updated the PR.

I moved the main logic to sea-orm as DatabaseExecutor, but I implemented SchemaManagerConnection in sea-orm-migration as a Newtype wrapper (struct) instead of a simple type alias.

Reason: If I used a type alias, I would have to implement sea_schema::Connection inside sea-orm core. This would force sea-orm core to depend on sea-schema, causing longer build times and unnecessary dependencies for users who don't use migrations.

Using the Newtype pattern keeps sea-orm core clean and lightweight, while strictly satisfying the Orphan Rule.

However, if you have a better idea or prefer a different approach, please let me know.

tyt2y3 commented on sea-orm-migration/src/connection.rs:
I think we can remove this from sea-orm-migration completely, and only re-export it? like 
```rust
pub use sea_orm::{IntoDatabaseExecutor as IntoSchemaManagerConnection, DatabaseExecutor as SchemaManagerConnection};
```

tyt2y3 reviewed (changes requested):
thank you for the PR!

I prefer to have only one type instead of having a second wrapper type.

> Separation of Concerns: Schema management logic (like get_schema_builder) belongs to the migration domain (DDL), not the core query execution domain (DML). The Newtype pattern allows us to keep this specific logic strictly within sea-orm-migration.

I agree, this is already achived in sea-orm with feature flags

> Dependency Hygiene: If I used a type alias, I would have to implement sea_schema::Connection inside sea-orm core. 
This would force sea-orm core to depend on sea-schema, causing longer build times for users who don't use migrations.

sea-schema is already a dependency of sea-orm, and I am tempted to merge it into sea-orm, then we don't have to have this shim. (don't worry, it's not happening right now)
this is gated behind the feature flag `schema-sync`. `sea-orm-migration` turn this flag on here https://github.com/SeaQL/sea-orm/blob/6e68ccd72f97548d92dc36fbcf219cb7db85ced1/sea-orm-migration/Cargo.toml#L27
